### PR TITLE
Reconcile Streaming Processors

### LINKED ISSUE #66: implement controllers for the streaming runtime

### PR DESCRIPTION
**Pull request by ericbottard**
Fixes #66

### CODE REVIEW

File: pkg/apis/streaming/v1alpha1/processor_lifecycle.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+func (ps *ProcessorStatus) MarkFunctionNotFound(name string) {
+	processorCondSet.Manage(ps).MarkFalse(ProcessorConditionFunctionReady, "NotFound",
+		"Unable to find function %q.", name)
+}
+
+func (ps *ProcessorStatus) PropagateFunctionStatus(fs *buildv1alpha1.FunctionStatus) {
+	ps.FunctionImage = fs.LatestImage
+	ps.FunctionImage = "ericbottard/fn" // TODO remove once build ctrl exists
```
Comment (scothis) [inline pkg/apis/streaming/v1alpha1/processor_lifecycle.go]: remove it now. You can create Function resources with `.status.latestImage` set for testing

File: pkg/apis/streaming/v1alpha1/processor_lifecycle.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+	if ps.FunctionImage == "" {
+		processorCondSet.Manage(ps).MarkFalse(ProcessorConditionFunctionReady, "NoImage",
+			"Function has no latestImage")
+	} else {
+		processorCondSet.Manage(ps).MarkTrue(ProcessorConditionFunctionReady)
+	}
+}
+
+func (ps *ProcessorStatus) MarkDeploymentNotOwned(name string) {
```
Comment (scothis) [inline pkg/apis/streaming/v1alpha1/processor_lifecycle.go]: ProcessorStatus has a DeploymentName field we can use for the message

```suggestion
func (ps *ProcessorStatus) MarkDeploymentNotOwned() {
```

File: pkg/apis/streaming/v1alpha1/processor_lifecycle.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	processorCondSet.Manage(ps).MarkFalse(ProcessorConditionDeploymentReady, "NotOwned",
+		"There is an existing Deployment %q that we do not own.", name)
+}
+
+func (ps *ProcessorStatus) MarkScaledObjectNotOwned(name string) {
```
Comment (scothis) [inline pkg/apis/streaming/v1alpha1/processor_lifecycle.go]: ProcessorStatus has a ScaledObjectName field we can use for the message

```suggestion
func (ps *ProcessorStatus) MarkScaledObjectNotOwned() {
```

File: pkg/apis/streaming/v1alpha1/processor_lifecycle.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	processorCondSet.Manage(ps).MarkFalse(ProcessorConditionScaledObjectReady, "NotOwned",
+		"There is an existing ScaledObject %q that we do not own.", name)
+}
+
+func (ps *ProcessorStatus) PropagateDeploymentStatus(ds *appsv1.DeploymentStatus) {
```
Comment (scothis) [inline pkg/apis/streaming/v1alpha1/processor_lifecycle.go]: fyi: propagating a valid status without an intermediate failed condition is a bit more complex than what is here. See https://github.com/projectriff/system/blob/master/pkg/apis/core/v1alpha1/deployer_lifecycle.go#L62-L88

File: pkg/controllers/streaming/processor_controller.go [hunk 5]
```diff
@@ -18,36 +18,506 @@ package streaming
...
+	Tracker   tracker.Tracker
+	Namespace string
 }
 
 // +kubebuilder:rbac:groups=streaming.projectriff.io,resources=processors,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=streaming.projectriff.io,resources=processors/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=build.projectriff.io,resources=functions,verbs=get;watch
+// +kubebuilder:rbac:groups=streaming.projectriff.io,resources=streams,verbs=get;watch
+// +kubebuilder:rbac:groups=keda.k8s.io,resources=scaledobject,verbs=get;watch
```
Comment (scothis) [inline pkg/controllers/streaming/processor_controller.go]: you need to create scaled objects, right? It should also move up to be next to deployment.

```suggestion
// +kubebuilder:rbac:groups=keda.k8s.io,resources=scaledobject,verbs= get;list;watch;create;update;patch;delete
```

File: pkg/controllers/streaming/processor_controller.go [hunk 6]
```diff
@@ -18,36 +18,506 @@ package streaming
...
+	}
+	return result, err
+}
+
+func (r *ProcessorReconciler) reconcile(ctx context.Context, logger logr.Logger, processor *streamingv1alpha1.Processor) (ctrl.Result, error) {
+	if processor.GetDeletionTimestamp() != nil {
+		return ctrl.Result{}, nil
+	}
+
+	// We may be reading a version of the object that was stored at an older version
```
Comment (scothis) [inline pkg/controllers/streaming/processor_controller.go]: for consistency with other controllers, move this line through `InitializeConditions()` out to the main Reconcile method just below `processor := original.DeepCopy()`.

File: pkg/controllers/streaming/processor_controller.go [hunk 7]
```diff
@@ -18,36 +18,506 @@ package streaming
...
+	Log       logr.Logger
+	Scheme    *runtime.Scheme
+	Tracker   tracker.Tracker
+	Namespace string
 }
 
 // +kubebuilder:rbac:groups=streaming.projectriff.io,resources=processors,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=streaming.projectriff.io,resources=processors/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=build.projectriff.io,resources=functions,verbs=get;watch
```
Comment (scothis) [inline pkg/controllers/streaming/processor_controller.go]: `get;list;watch` is a more common permission set for read-only access

**Comment by ericbottard** [comment]
Merged as ddde6ac

**Pull request closed without merging**